src/libcrun/criu.c

```diff
@@ -0,0 +1,147 @@
+   * is a minimum CRIU version listed and so far it is good enough.
+   *
+   * The CRIU library also does not yet have an interface to CRIU
+   * the version of the binary. Right now it is only possible to
+   * query the version of the library via defines during buildtime.
+   *
+   * The whole CRIU library setup works this way, that the library
+   * is only a wrapper around RPC calls to the actual library. So
+   * if CRIU is updated and the SO of the library does not change,
+   * and crun is not rebuilt against the newer version, the version
+   * is still returning the values during buildtime and not from
+   * the actual running CRIU binary. The RPC interface between the
+   * library will not break, so no reason to worry, but it is not
+   * possible to detect (via the library) which CRIU version is
+   * actually being used. This needs to be added to CRIU upstream. */
+
+  ret = criu_init_opts ();
+  if (UNLIKELY (ret < 0))
+    return crun_make_error (err, 0, "CRIU init failed with %d\n", ret);
+
+  if (UNLIKELY (cr_options->image_path == NULL))
+    return crun_make_error (err, 0, "image path not set\n");
+
+  ret = mkdir (cr_options->image_path, 0700);
+  if (UNLIKELY ((ret == -1) && (errno != EEXIST)))
+    return crun_make_error (err, 0,
+                            "error creating checkpoint directory %s\n",
+                            cr_options->image_path);
+
+  image_fd = open (cr_options->image_path, O_DIRECTORY);
+  if (UNLIKELY (image_fd == -1))
+    return crun_make_error (err, 0, "error opening checkpoint directory %s\n",
+                            cr_options->image_path);
+
+  criu_set_images_dir_fd (image_fd);
+
+  /* work_dir is the place CRIU will put its logfiles. If not explicitly set,
+   * CRIU will put the logfiles into the images_dir from above. No need for
+   * crun to set it if the user has not selected a specific directory. */
+  if (cr_options->work_path != NULL)
+    {
+      work_fd = open (cr_options->work_path, O_DIRECTORY);
+      if (UNLIKELY (work_fd == -1))
+        return crun_make_error (err, 0,
+                                "error opening CRIU work directory %s\n",
+                                cr_options->work_path);
+
+      criu_set_work_dir_fd (work_fd);
+    }
+  else
+    {
+      /* This is only for the error message later. */
+      cr_options->work_path = cr_options->image_path;
+    }
+
+  /* The main process of the container is the process CRIU will checkpoint
+   * and all of its children. */
+  criu_set_pid (status->pid);
+
+  ret = xasprintf (&path, "%s/%s", status->bundle, status->rootfs);
+  if (UNLIKELY (ret == -1))
+    libcrun_fail_with_error (0, "xasprintf failed");
+
+  ret = criu_set_root (path);
+  if (UNLIKELY (ret != 0))
+    return crun_make_error (err, 0, "error setting CRIU root to %s\n", path);
+
+  /* Tell CRIU about external bind mounts. */
+  for (i = 0; i < def->mounts_len; i++)
+    {
+      if (strcmp (def->mounts[i]->type, "bind") == 0)
```

Comment:
the correct way to check whether it is a bind mount is to check the options for "bind" or "rbind".  In any case it is something we can address later.

Reply:
The right thing to do would be something like this line from linux.c?
```
if (strcmp (def->mounts[i]->options[j], "bind") == 0 || strcmp (def->mounts[i]->options[j], "rbind") == 0)
```

Reply:
yes thanks, that will work